### PR TITLE
Add Gazebo Rover

### PR DESCRIPTION
**Pull request by AlexKlimaj**
Currently not working. Missing all sensor data.

Brought in @bperseghetti https://app.gazebosim.org/RudisLaboratories/fuel/models/MR-Buggy3

### CODE REVIEW

**Comment by psvssk** [comment]
Is there any update this

File: Tools/simulation/gz/models/rover/model.sdf [hunk 1]
```diff
@@ -0,0 +1,388 @@
+<sdf version="1.8">
+  <model name="rover">
+    <pose>0 0 .04 0 0 0</pose>
+    <self_collide>false</self_collide>
+    <static>false</static>
```
Comment (Jaeyoung-Lim) [inline Tools/simulation/gz/models/rover/model.sdf:5]: Could we include this directly from ignition fuel?

**Review by Jaeyoung-Lim — CHANGES_REQUESTED** [review]

File: Tools/simulation/gz/models/rover/model.sdf [hunk 2]
```diff
@@ -0,0 +1,388 @@
...
+    <link name="rover/Base">
```
Comment (psvssk) [inline Tools/simulation/gz/models/rover/model.sdf:6]: Currently, the imu sensor is not getting added to the simulation. 'rover/Base' has to be renamed to 'base_link'.

**Comment by psvssk** [comment]
Does anyone know how to map the actuators from PX4 to the GZ simulation? 
Currently, I have added the following lines to the airframe configuration. 


. ${R}etc/init.d/rc.rover_defaults

PX4_SIMULATOR=${PX4_SIMULATOR:=gz}
PX4_GZ_WORLD=${PX4_GZ_WORLD:=default}
PX4_SIM_MODEL=${PX4_SIM_MODEL:=rover}

param set-default SIM_GZ_EN 1

param set-default SENS_EN_GPSSIM 1
param set-default SENS_EN_BAROSIM 0
param set-default SENS_EN_MAGSIM 1

param set-default CBRK_AIRSPD_CHK 162128

param set-default CA_AIRFRAME 5
param set-default CA_R_REV 1

param set-default SIM_GZ_EC_FUNC1 201
param set-default SIM_GZ_EC_FUNC2 201
param set-default SIM_GZ_EC_FUNC6 101
param set-default SIM_GZ_EC_FUNC7 101

param set-default SIM_GZ_EC_MIN1 10
param set-default SIM_GZ_EC_MIN2 10
param set-default SIM_GZ_EC_MIN6 10
param set-default SIM_GZ_EC_MIN7 10

param set-default SIM_GZ_EC_MAX1 1000
param set-default SIM_GZ_EC_MAX2 1000
param set-default SIM_GZ_EC_MAX6 1000
param set-default SIM_GZ_EC_MAX7 1000

**Comment by bperseghetti** [comment]
@psvssk here is the one with the IMU and lidar: https://github.com/CogniPilot/mrbuggy3_simulator/tree/main/mrbuggy3_gz_resource/models/mrbuggy3

**Comment by bperseghetti** [comment]
> @AlexKlimaj @Jaeyoung-Lim @psvssk - this PR looks stale. Maybe this could help, I have a work in progress implementation of a lawn mower in Gazebo ("gz", Ignition). It runs at "make px4_sitl gz_lawnmower". Here are links to important files:
> 
> https://github.com/slgrobotics/PX4-Autopilot/blob/main/ROMFS/px4fmu_common/init.d-posix/airframes/5005_gz_lawnmower
> 
> https://github.com/slgrobotics/PX4-Autopilot/tree/main/Tools/simulation/gz/models/lawnmower
> 
> I used "JointPositionController" for wheels, but currently there is no reverse and I see no way to scale and offset input signal (for example, use servo range 1000 to 2000 with wheels stopped at 1500).

Are you running Ackermann steering? If you are, I would suggest using the steering only component (https://github.com/CogniPilot/mrbuggy3_simulator/blob/bc4aee28477dc3ba5467f935fbec91ecac4d07ec/mrbuggy3_gz_resource/models/mrbuggy3/model.sdf#L431) of it and then use the JointController to set the wheel angular rate (I think what can best be done is use the actuator message that already exists in sim and that we already use to send true (not normalized) velocity and position commands, you can make it use that by: https://github.com/CogniPilot/mrbuggy3_simulator/blob/bc4aee28477dc3ba5467f935fbec91ecac4d07ec/mrbuggy3_gz_resource/models/mrbuggy3/model.sdf#L425-L426C43

**Comment by slgrobotics** [comment]
@bperseghetti  - no, I have a gas engine driven full size zero-turn lawnmower, and each wheel is independently controlled by levers, moved by servos. So, it is essentially a differential drive rover. I don't want to have any module (like the gz diff drive) to be on the way between my commands (1000..1500..2000) and the wheels. That's why I am trying to use JointController plugin. 

Gazebo Classic handled it exactly as I wanted via "mavlink_interface" plugin, and that worked fine. I could just directly control "motors", setting scaling and PID values for each channel:
https://github.com/slgrobotics/PX4-Autopilot/blob/main/Tools_gazebo/r1_rover.sdf.jinja

Unfortunately, GZ JointController plugin is less sophisticated.

**Comment by bperseghetti** [comment]
> @bperseghetti - no, I have a gas engine driven full size zero-turn lawnmower, and each wheel is independently controlled by levers, moved by servos. So, it is essentially a differential drive rover. I don't want to have any module (like the gz diff drive) to be on the way between my commands (1000..1500..2000) and the wheels. That's why I am trying to use JointController plugin.
> 
> Gazebo Classic handled it exactly as I wanted via "mavlink_interface" plugin, and that worked fine. I could just directly control "motors", setting scaling and PID values for each channel: https://github.com/slgrobotics/PX4-Autopilot/blob/main/Tools_gazebo/r1_rover.sdf.jinja
> 
> Unfortunately, GZ JointController plugin is less sophisticated.

JointController takes in values of rad/sec which I would argue are better units to use since if trying to model PWM it is just normalized random non-sense anyhow that you would then have to remap to rad/sec.

You could write something to sit between that converts your PWM scale range into rad/sec and feeds that into the actuator message. Or use the PX4 normalized -1,1 output and scale it based on what you see with your vehicle in real world over PWM range but in rad/sec. Tachometer is a good thing to clock the wheels with.

**Comment by chfriedrich98** [comment]
@AlexKlimaj I'm going through old rover related issues/PRs and wanted to ask if it is fine to close this one?
The rover support has since been updated, deprecating the rover_pos_ctrl module and https://github.com/PX4/PX4-gazebo-models/pull/46 added a gazebo ackermann rover model that is based on this model (different mehses).

**Pull request closed without merging**